### ISSUE TITLE
Kronecker-factored matrices

### ISSUE THREAD

**Issue by willtebbutt** [JuliaGaussianProcesses/Stheno.jl#6]
Consider a kernel of the form
```julia
(k)(x::Vector, x′::Vector) = prod(map(k.ks, x, x′))
```
where `k.ks` is a collection of kernels which can be applied to single-dimensions, and the dimensionality of `x` and `x′`, and the length of `k.ks` is `D`.
If we wish to evaluate the covariance matrix associated with a matrix of inputs which can be represented as a Cartesian product:
```julia
X = IterTools.product(x1, ..., xD)
```
where each `xd` is a vector, then `cov(k, X)` can be represented as the Kronecker product of a set of `D` matrices, the `d`th of which is `cov(k.ks[d], xd)`. If this structure is present then the asymptotic complexity of inference is greatly reduced.

- Details of this can be found in [1].
- There was work at NIPS 2017 in the AABI workshop which maybe relates to this work - see [2].
- There is no obvious reason why one shouldn't be able to exploit matrices which admit partial Kronecker factorisations. I'm not sure what / if there is relevant work is here though.
- Perhaps there is something in the tensor decomposition / tensor train field / line-of-work (I'm not even sure what to call this) that can speak to the above. This video looks related: https://www.youtube.com/watch?v=CrSFne4h3Pk

To achieve efficient inference here, it will be necessary to implement a Kronecker-factorised matrix type which has appropriate efficient representations.

[1] - Saatçi, Yunus. Scalable inference for structured Gaussian process models. Diss. University of Cambridge, 2012.
[2] - Evans, Trefor W., and Prasanth B. Nair. "Scalable Gaussian Processes with Grid-Structured Eigenfunctions (GP-GRIEF)."

**Comment by willtebbutt** [comment]
[Kronecker.jl](https://github.com/MichielStock/Kronecker.jl) is a thing

**Comment by MichielStock** [comment]
This is a large body of research using Kronecker factorizations in GP/kernel ridge regression, e.g. [GPatt](https://arxiv.org/pdf/1310.5288.pdf) and [Kronecker ridge regression](https://www.mitpressjournals.org/doi/abs/10.1162/neco_a_01096?journalCode=neco). Would love to discuss this or contribute!

**Comment by willtebbutt** [comment]
Would be great to both discuss and collaborate :) Are you around for the hackathon tomorrow?

**Comment by MichielStock** [comment]
Yes! I am here the whole Hackathon. Would like to get involved 🙂
